File: src/rag.py
"""RAG-Kern: Retrieval aus dem Vektorspeicher + Antwort ausschließlich
auf Basis der gefundenen BookStack-Inhalte.
"""
from __future__ import annotations

from dataclasses import dataclass
from typing import Iterator

from config import Config
from src.ollama_client import OllamaClient
from src.vectorstore import Retrieved, VectorStore

SYSTEM_PROMPT = """\
Du bist der Wiki-Assistent und beantwortest Fragen AUSSCHLIESSLICH auf Basis \
der bereitgestellten Auszüge aus dem BookStack-Wiki (KONTEXT).

Regeln:
- Nutze NUR Informationen aus dem KONTEXT. Verwende KEIN sonstiges Vorwissen.
- Wenn die Antwort nicht im KONTEXT steht, sage klar: "Das steht nicht im Wiki."
- Erfinde nichts und rate nicht.
- Antworte in der Sprache der Frage, präzise und knapp.
- Verweise wo sinnvoll auf die Quelle über die Nummer in eckigen Klammern, z.B. [1].
"""

NO_CONTEXT_MSG = "Dazu finde ich nichts im Wiki."


@dataclass
class Answer:
    text: str
    sources: list[Retrieved]


def _build_context(chunks: list[Retrieved]) -> str:
    blocks = []
    for i, ch in enumerate(chunks, start=1):
        meta = ch.metadata
        loc = " > ".join(
            p for p in [meta.get("book_name"), meta.get("page_name")] if p
        )
        blocks.append(f"[{i}] Quelle: {loc}\n{ch.text}")
    return "\n\n---\n\n".join(blocks)


def _format_sources(chunks: list[Retrieved]) -> str:
    seen: dict[int, dict] = {}
    order: list[int] = []
    for ch in chunks:
        pid = ch.metadata.get("page_id")
        if pid not in seen:
            seen[pid] = ch.metadata
            order.append(pid)
    lines = []
    for i, pid in enumerate(order, start=1):
        meta = seen[pid]
        loc = " > ".join(
            p for p in [meta.get("book_name"), meta.get("page_name")] if p
        )
        url = meta.get("url", "")
        lines.append(f"[{i}] {loc} – {url}" if url else f"[{i}] {loc}")
    return "\n".join(lines)


class RagEngine:
    def __init__(self, cfg: Config):
        self.cfg = cfg
        self.ollama = OllamaClient(cfg.ollama_url)
        self.store = VectorStore(cfg.chroma_dir, cfg.chroma_collection)

    def retrieve(self, question: str) -> list[Retrieved]:
        q_emb = self.ollama.embed(self.cfg.ollama_embed_model, question)
        hits = self.store.query(q_emb, self.cfg.retrieval_top_k)
        # Nach Distanzschwelle filtern (irrelevante Treffer verwerfen)
        return [h for h in hits if h.distance <= self.cfg.retrieval_max_distance]

    def _user_prompt(self, question: str, chunks: list[Retrieved]) -> str:
        context = _build_context(chunks)
        return f"KONTEXT:\n{context}\n\nFRAGE: {question}\n\nAntwort:"

    def answer(self, question: str) -> Answer:
        chunks = self.retrieve(question)
        if not chunks:
            return Answer(text=NO_CONTEXT_MSG, sources=[])
        text = self.ollama.chat(
            self.cfg.ollama_llm_model,
            SYSTEM_PROMPT,
            self._user_prompt(question, chunks),
        )
        return Answer(text=text, sources=chunks)

    def answer_stream(self, question: str) -> tuple[Iterator[str], list[Retrieved]]:
        """Gibt (Token-Stream, Quellen) zurück. Bei fehlendem Kontext ein Ein-Token-Stream."""
        chunks = self.retrieve(question)
        if not chunks:
            return iter([NO_CONTEXT_MSG]), []
        stream = self.ollama.chat_stream(
            self.cfg.ollama_llm_model,
            SYSTEM_PROMPT,
            self._user_prompt(question, chunks),
        )
        return stream, chunks

    @staticmethod
    def format_sources(chunks: list[Retrieved]) -> str:
        return _format_sources(chunks)

File: internal/rag/rag.go
// Package rag verbindet Retrieval (Vektorspeicher) mit der Antwortgenerierung.
// Es wird ausschließlich der abgerufene BookStack-Kontext genutzt.
package rag

import (
	"context"
	"fmt"
	"strings"

	"wiki-rog/internal/config"
	"wiki-rog/internal/ollama"
	"wiki-rog/internal/store"
)

const SystemPrompt = `Du bist der Wiki-Assistent und beantwortest Fragen AUSSCHLIESSLICH auf Basis der bereitgestellten Auszüge aus dem BookStack-Wiki (KONTEXT).

Regeln:
- Nutze NUR Informationen aus dem KONTEXT. Verwende KEIN sonstiges Vorwissen.
- Wenn die Antwort nicht im KONTEXT steht, sage klar: "Das steht nicht im Wiki."
- Erfinde nichts und rate nicht.
- Antworte in der Sprache der Frage, präzise und knapp.
- Verweise wo sinnvoll auf die Quelle über die Nummer in eckigen Klammern, z.B. [1].`

const NoContextMsg = "Dazu finde ich nichts im Wiki."

type Engine struct {
	cfg    config.Config
	ollama *ollama.Client
	store  store.Store
}

func New(cfg config.Config, oc *ollama.Client, st store.Store) *Engine {
	return &Engine{cfg: cfg, ollama: oc, store: st}
}

// Retrieve holt die relevanten Chunks (nach Distanzschwelle gefiltert).
func (e *Engine) Retrieve(ctx context.Context, question string) ([]store.Result, error) {
	emb, err := e.ollama.Embed(ctx, e.cfg.OllamaEmbedModel, question)
	if err != nil {
		return nil, err
	}
	hits, err := e.store.Query(ctx, emb, e.cfg.RetrievalTopK)
	if err != nil {
		return nil, err
	}
	out := hits[:0]
	for _, h := range hits {
		if h.Distance <= e.cfg.RetrievalMaxDistance {
			out = append(out, h)
		}
	}
	return out, nil
}

func buildContext(chunks []store.Result) string {
	var b strings.Builder
	for i, ch := range chunks {
		loc := joinNonEmpty(" > ", ch.Meta.BookName, ch.Meta.PageName)
		if i > 0 {
			b.WriteString("\n\n---\n\n")
		}
		fmt.Fprintf(&b, "[%d] Quelle: %s\n%s", i+1, loc, ch.Document)
	}
	return b.String()
}

func (e *Engine) userPrompt(question string, chunks []store.Result) string {
	return fmt.Sprintf("KONTEXT:\n%s\n\nFRAGE: %s\n\nAntwort:", buildContext(chunks), question)
}

// Answer liefert die vollständige Antwort inkl. Quellen.
func (e *Engine) Answer(ctx context.Context, question string) (string, []store.Result, error) {
	chunks, err := e.Retrieve(ctx, question)
	if err != nil {
		return "", nil, err
	}
	if len(chunks) == 0 {
		return NoContextMsg, nil, nil
	}
	text, err := e.ollama.Chat(ctx, e.cfg.OllamaLLMModel, SystemPrompt, e.userPrompt(question, chunks))
	if err != nil {
		return "", nil, err
	}
	return text, chunks, nil
}

// AnswerStream streamt die Antwort token-weise über onToken und gibt die Quellen zurück.
func (e *Engine) AnswerStream(ctx context.Context, question string, onToken func(string)) ([]store.Result, error) {
	chunks, err := e.Retrieve(ctx, question)
	if err != nil {
		return nil, err
	}
	if len(chunks) == 0 {
		onToken(NoContextMsg)
		return nil, nil
	}
	err = e.ollama.ChatStream(ctx, e.cfg.OllamaLLMModel, SystemPrompt, e.userPrompt(question, chunks), onToken)
	return chunks, err
}

// FormatSources erzeugt eine nummerierte Quellenliste (dedupliziert je Seite).
func FormatSources(chunks []store.Result) string {
	seen := map[int]bool{}
	var lines []string
	n := 0
	for _, ch := range chunks {
		if seen[ch.Meta.PageID] {
			continue
		}
		seen[ch.Meta.PageID] = true
		n++
		loc := joinNonEmpty(" > ", ch.Meta.BookName, ch.Meta.PageName)
		if ch.Meta.URL != "" {
			lines = append(lines, fmt.Sprintf("[%d] %s – %s", n, loc, ch.Meta.URL))
		} else {
			lines = append(lines, fmt.Sprintf("[%d] %s", n, loc))
		}
	}
	return strings.Join(lines, "\n")
}

func joinNonEmpty(sep string, parts ...string) string {
	var out []string
	for _, p := range parts {
		if strings.TrimSpace(p) != "" {
			out = append(out, p)
		}
	}
	return strings.Join(out, sep)
}
